### DISCUSSION TITLE
How to use extension functions

### DISCUSSION THREAD

**Discussion by ahka01** [comunica/comunica#856] [Q&A] [answered]
Question:

I'm trying to have the example provided in the section: https://comunica.dev/docs/query/advanced/extension_functions/ working, but still no success.

```javascript
const newEngine = require('@comunica/actor-init-sparql').newEngine;
const myEngine = newEngine();
const df = require('rdf-data-factory');

async function main() {

    const DF = new df.DataFactory();

    const result = await myEngine.query(`
    PREFIX func: <http://example.org/functions#>
    SELECT ?caps WHERE {
    ?s ?p ?o.
    BIND (func:to-upper-case(?o) AS ?caps)
    }
`, {
        sources: ['https://www.rubensworks.net/'],
        extensionFunctions: {
            'http://example.org/functions#to-upper-case'(args) {
                const arg = args[0];
                if (arg.termType === 'Literal' && arg.datatype.value === 'http://www.w3.org/2001/XMLSchema#string') {
                    return DF.literal(arg.value.toUpperCase(), arg.datatype);
                }
                return arg;
            },
        },
    });
}
main();

```

The following error is being returned:

`
(node:27870) UnhandledPromiseRejectionWarning: Error: All actors rejected their test in https://linkedsoftwaredependencies.org/bundles/npm/@comunica/actor-init-sparql/^1.0.0/config/sets/sparql-queryoperators.json#mediatorQueryOperation
Unknown named operator: '"http://example.org/functions#to-upper-case"'
https://linkedsoftwaredependencies.org/bundles/npm/@comunica/actor-init-sparql/^1.0.0/config/sets/resolve-sparql.json#mySparqlEndpointResolver requires a single source with a 'sparql' endpoint to be present in the context.
    at MediatorNumber.mediateWith (/Users/ahmadkadi/Documents/Home/DFKI/GeoBox/Others/TrainingFolder/NodeJS/20200128_communica/node_modules/@comunica/mediator-number/lib/MediatorNumber.js:54:19)
    at process._tickCallback (internal/process/next_tick.js:68:7)
    at Function.Module.runMain (internal/modules/cjs/loader.js:832:11)
    at startup (internal/bootstrap/node.js:283:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:622:3)
(node:27870) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 3)
(node:27870) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
`
Thanks for your support

**Comment by rubensworks** [comment, marked as answer]
Are you sure you're using the latest version of Comunica? (1.22.0)

**Reply by ahka01** [reply]
Thanks for the hint. I had an earlier version. My bad!

**Comment by ahka01** [comment]
Hello again,

I have different issue regarding using extension function inside FILTER in a SPARQL query.

Starting from the following dummy example:

```javascript
const newEngine = require('@comunica/actor-init-sparql').newEngine;
const myEngine = newEngine();
const df = require('rdf-data-factory');
const N3 = require('n3');
const store = new N3.Store();
const parser = new N3.Parser();
const DF = new df.DataFactory();

async function call() {

  let startDate = '2015-09-03T08:43:23.000Z';
  let file = `
  @prefix m: <http://www.example.de/m/> .
  @prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
  m:a m:dateFrom "2014-07-03T08:43:23.000Z"^^xsd:dateTime .
  m:b m:dateFrom "2015-07-03T08:43:23.000Z"^^xsd:dateTime .
  m:c m:dateFrom "2016-07-03T08:43:23.000Z"^^xsd:dateTime .
  `;
  parser.parse(file,
    (error, quad, prefixes) => {
      if (quad)
        store.addQuad(quad);
      else 
        return Promise.resolve();
    });

      const query  =`
      PREFIX func: <http://example.org/functions#>
      PREFIX m: <http://www.example.de/m/> 
      SELECT ?s ?from 
      WHERE { ?s m:dateFrom ?from . FILTER(func:getTimeLessThan(?from, '${startDate}'))}`;
  // Initiate the update
   const result = await myEngine.query(
    query , {
    sources: [store],
    extensionFunctions: {
      'http://example.org/functions#getTimeLessThan'(args) {
        const timeIntersectValue = timeLessThan(args[0].value, args[1].value);
        return DF.literal(timeIntersectValue, DF.namedNode('http://www.w3.org/2001/XMLSchema#boolean'));
      },
    },
   });
  console.log((await result.bindings()).length);
}

function timeLessThan(dataParam, upperTimeLimit) {
  const timeIntersect = ((new Date(dataParam)).getTime() < (new Date(upperTimeLimit)).getTime());
  console.log(timeIntersect);
  return timeIntersect;
}

call();

```

I'm expecting to get two results; however nothing is returned. The length of the returned result array is 0.
This is the prints of the log statements I added 
`true
true
false
0`

Is my expectation to get two rows of results is wrong here? Or what I'm missing to add?

Thanks for the help
Best regards

**Reply by rubensworks** [reply]
Not sure, but it might be related to the fact that you're passing a boolean to `DF.literal`, while a string is required. So converting that boolean to `"true"` or `"false"` might do the trick.

**Reply by ahka01** [reply]
Thanks a lot. It works now
